File: src/backend/api/schemas/reservation.py
"""
Pydantic schema models for GPU reservation management in the Provocative Cloud platform,
with integrated environmental impact tracking and carbon capture metrics.
"""

from datetime import datetime, timedelta
from decimal import Decimal
from typing import Dict, Optional
from uuid import UUID

from pydantic import BaseModel, Field, validator

from api.schemas.gpu import GPUBase
from api.schemas.billing import PaymentBase

# Reservation status choices
RESERVATION_STATUS_CHOICES = ('pending', 'active', 'completed', 'cancelled', 'failed')
MIN_RENTAL_HOURS = 1
MAX_RENTAL_HOURS = 720  # 30 days
DEPLOYMENT_STATUS_CHOICES = ('pending', 'provisioning', 'ready', 'error')
COOLING_STATUS_CHOICES = ('optimal', 'degraded', 'critical')
CARBON_PREFERENCE_CHOICES = ('standard', 'eco_friendly', 'maximum_efficiency')

class ReservationBase(BaseModel):
    """Base schema for GPU reservation data with environmental metrics."""
    id: Optional[UUID] = Field(default=None)
    user_id: UUID = Field(..., description="User making the reservation")
    gpu_id: UUID = Field(..., description="GPU being reserved")
    start_time: datetime = Field(..., description="Reservation start time")
    end_time: datetime = Field(..., description="Reservation end time")
    auto_renew: bool = Field(default=False, description="Auto-renewal flag")
    status: str = Field(default='pending', description="Reservation status")
    created_at: datetime = Field(default_factory=datetime.utcnow)
    environmental_impact: Dict[str, float] = Field(
        default_factory=dict,
        description="Environmental impact metrics"
    )
    cooling_status: str = Field(
        default='optimal',
        description="Cooling system status"
    )

    @validator('end_time')
    def validate_times(cls, end_time: datetime, values: Dict) -> datetime:
        """Validates reservation time period with cooling system checks."""
        start_time = values.get('start_time')
        if not start_time:
            raise ValueError("start_time is required")

        # Validate time order
        if end_time <= start_time:
            raise ValueError("end_time must be after start_time")

        # Calculate duration
        duration = (end_time - start_time).total_seconds() / 3600

        # Validate rental period
        if duration < MIN_RENTAL_HOURS:
            raise ValueError(f"Minimum rental period is {MIN_RENTAL_HOURS} hour")
        if duration > MAX_RENTAL_HOURS:
            raise ValueError(f"Maximum rental period is {MAX_RENTAL_HOURS} hours")

        return end_time

    @validator('environmental_impact')
    def validate_environmental_impact(cls, value: Dict[str, float]) -> Dict[str, float]:
        """Validates environmental impact metrics."""
        required_metrics = {
            'co2_captured_kg',
            'power_usage_kwh',
            'cooling_efficiency',
            'carbon_offset_kg'
        }

        # Validate required metrics
        if not all(metric in value for metric in required_metrics):
            raise ValueError(f"Required environmental metrics: {', '.join(required_metrics)}")

        # Validate metric ranges
        if not 0 <= value['cooling_efficiency'] <= 1:
            raise ValueError("Cooling efficiency must be between 0 and 1")

        if value['co2_captured_kg'] < 0:
            raise ValueError("CO2 capture cannot be negative")

        if value['power_usage_kwh'] < 0:
            raise ValueError("Power usage cannot be negative")

        if value['carbon_offset_kg'] < 0:
            raise ValueError("Carbon offset cannot be negative")

        return value

    @validator('cooling_status')
    def validate_cooling_status(cls, value: str) -> str:
        """Validates cooling system status."""
        if value not in COOLING_STATUS_CHOICES:
            raise ValueError(f"Cooling status must be one of: {', '.join(COOLING_STATUS_CHOICES)}")
        return value

class ReservationCreate(BaseModel):
    """Schema for creating new GPU reservations with environmental preferences."""
    user_id: UUID = Field(..., description="User making the reservation")
    gpu_id: UUID = Field(..., description="GPU to reserve")
    start_time: datetime = Field(..., description="Desired start time")
    duration_hours: int = Field(
        ...,
        ge=MIN_RENTAL_HOURS,
        le=MAX_RENTAL_HOURS,
        description="Rental duration in hours"
    )
    auto_renew: bool = Field(default=False, description="Auto-renewal flag")
    carbon_preference: str = Field(
        default='standard',
        description="Carbon capture preference"
    )

    @validator('carbon_preference')
    def validate_carbon_preference(cls, value: str) -> str:
        """Validates carbon preference selection."""
        if value not in CARBON_PREFERENCE_CHOICES:
            raise ValueError(f"Carbon preference must be one of: {', '.join(CARBON_PREFERENCE_CHOICES)}")
        return value

    @validator('start_time')
    def validate_start_time(cls, value: datetime) -> datetime:
        """Validates reservation start time."""
        now = datetime.utcnow()
        if value < now:
            raise ValueError("Start time cannot be in the past")
        return value

class ReservationResponse(BaseModel):
    """Schema for reservation API responses with environmental metrics."""
    reservation: ReservationBase
    gpu: GPUBase
    payment: PaymentBase
    deployment_status: str = Field(default='pending')
    ssh_connection_string: Optional[str] = None
    jupyter_url: Optional[str] = None
    environmental_metrics: Dict[str, float] = Field(
        default_factory=dict,
        description="Real-time environmental metrics"
    )
    cooling_efficiency: float = Field(
        default=1.0,
        ge=0.0,
        le=1.0,
        description="Current cooling system efficiency"
    )
    carbon_impact_report: Dict[str, Any] = Field(
        default_factory=dict,
        description="Detailed carbon impact analysis"
    )

    @validator('deployment_status')
    def validate_deployment_status(cls, value: str) -> str:
        """Validates deployment status."""
        if value not in DEPLOYMENT_STATUS_CHOICES:
            raise ValueError(f"Deployment status must be one of: {', '.join(DEPLOYMENT_STATUS_CHOICES)}")
        return value

    @validator('environmental_metrics')
    def validate_environmental_metrics(cls, value: Dict[str, float]) -> Dict[str, float]:
        """Validates real-time environmental metrics."""
        required_metrics = {
            'current_power_usage_watts',
            'temperature_celsius',
            'co2_capture_rate_kgh',
            'cooling_power_usage_watts'
        }

        if not all(metric in value for metric in required_metrics):
            raise ValueError(f"Required real-time metrics: {', '.join(required_metrics)}")

        return value

File: src/web/src/types/reservation.ts
/**
 * TypeScript type definitions for GPU reservation management
 * Provides type-safe interfaces for creating, updating, and tracking GPU rentals
 * @version 1.0.0
 */

import { UUID, Timestamp } from './common';
import { GPU } from './gpu';
import { GPUMetrics } from './metrics';

/**
 * Enum for reservation status with strict state transitions
 */
export enum ReservationStatus {
  PENDING = 'pending',
  ACTIVE = 'active',
  COMPLETED = 'completed',
  CANCELLED = 'cancelled',
  FAILED = 'failed'
}

/**
 * Enum for deployment status with validated state transitions
 */
export enum DeploymentStatus {
  PENDING = 'pending',
  PROVISIONING = 'provisioning',
  READY = 'ready',
  ERROR = 'error'
}

/**
 * Main interface for GPU reservations with cost and environmental impact tracking
 */
export interface Reservation {
  readonly id: UUID;
  readonly user_id: UUID;
  readonly gpu_id: UUID;
  readonly start_time: Timestamp;
  readonly end_time: Timestamp;
  /**
   * Flag for automatic renewal of reservation
   */
  readonly auto_renew: boolean;
  readonly status: ReservationStatus;
  readonly created_at: Timestamp;
  /**
   * Total cost of reservation in USD
   * @precision 2 decimal places
   */
  readonly total_cost: number;
  /**
   * Amount of CO2 captured during reservation in kg
   * @precision 2 decimal places
   */
  readonly carbon_offset: number;
}

/**
 * Interface for creating new reservations with deployment preferences
 */
export interface ReservationCreate {
  readonly user_id: UUID;
  readonly gpu_id: UUID;
  readonly start_time: Timestamp;
  /**
   * Duration of reservation in hours
   * @minimum MIN_RENTAL_HOURS
   * @maximum MAX_RENTAL_HOURS
   */
  readonly duration_hours: number;
  readonly auto_renew: boolean;
  /**
   * Type of deployment environment requested
   */
  readonly deployment_type: 'ssh' | 'jupyter' | 'docker';
}

/**
 * Interface for detailed reservation information including GPU and deployment details
 */
export interface ReservationDetails {
  readonly reservation: Reservation;
  readonly gpu: GPU;
  readonly deployment_status: DeploymentStatus;
  /**
   * SSH connection string if deployment_type is 'ssh'
   */
  readonly ssh_connection_string: string | null;
  /**
   * Jupyter notebook URL if deployment_type is 'jupyter'
   */
  readonly jupyter_url: string | null;
  /**
   * Real-time GPU performance metrics
   */
  readonly metrics: GPUMetrics;
  /**
   * Deployment progress and error logs
   */
  readonly deployment_logs: string[];
}

/**
 * Global constants for reservation management
 */
export const MIN_RENTAL_HOURS = 1;
export const MAX_RENTAL_HOURS = 720; // 30 days
export const DEPLOYMENT_TIMEOUT_SECONDS = 300;
export const AUTO_RENEW_THRESHOLD_HOURS = 2;

/**
 * Type guard to validate reservation creation parameters
 */
export function isValidReservationCreate(params: unknown): params is ReservationCreate {
  if (!params || typeof params !== 'object') {
    return false;
  }

  const p = params as ReservationCreate;

  return (
    typeof p.user_id === 'string' &&
    typeof p.gpu_id === 'string' &&
    typeof p.start_time === 'number' &&
    typeof p.duration_hours === 'number' &&
    p.duration_hours >= MIN_RENTAL_HOURS &&
    p.duration_hours <= MAX_RENTAL_HOURS &&
    typeof p.auto_renew === 'boolean' &&
    ['ssh', 'jupyter', 'docker'].includes(p.deployment_type)
  );
}

/**
 * Type guard to validate reservation status transitions
 */
export function isValidStatusTransition(
  currentStatus: ReservationStatus,
  newStatus: ReservationStatus
): boolean {
  const validTransitions: Record<ReservationStatus, ReservationStatus[]> = {
    [ReservationStatus.PENDING]: [
      ReservationStatus.ACTIVE,
      ReservationStatus.FAILED,
      ReservationStatus.CANCELLED
    ],
    [ReservationStatus.ACTIVE]: [
      ReservationStatus.COMPLETED,
      ReservationStatus.CANCELLED,
      ReservationStatus.FAILED
    ],
    [ReservationStatus.COMPLETED]: [],
    [ReservationStatus.CANCELLED]: [],
    [ReservationStatus.FAILED]: []
  };

  return validTransitions[currentStatus]?.includes(newStatus) ?? false;
}

/**
 * Type guard to validate deployment status transitions
 */
export function isValidDeploymentTransition(
  currentStatus: DeploymentStatus,
  newStatus: DeploymentStatus
): boolean {
  const validTransitions: Record<DeploymentStatus, DeploymentStatus[]> = {
    [DeploymentStatus.PENDING]: [
      DeploymentStatus.PROVISIONING,
      DeploymentStatus.ERROR
    ],
    [DeploymentStatus.PROVISIONING]: [
      DeploymentStatus.READY,
      DeploymentStatus.ERROR
    ],
    [DeploymentStatus.READY]: [
      DeploymentStatus.ERROR
    ],
    [DeploymentStatus.ERROR]: [
      DeploymentStatus.PENDING
    ]
  };

  return validTransitions[currentStatus]?.includes(newStatus) ?? false;
}